### PR TITLE
Raise error for non existing config options

### PR DESCRIPTION
**Pull request by philippmwirth**
# Raise error for non existing config options

Do basic validation of configuration options such that there's no silent failure when an argument is misspelled (e.g. `relevant_filenames` vs `relevant_filenames_file`.

A few examples:
___ 
```python
client.schedule_compute_worker_run(
    worker_config={
        "relevant_filenames": "my_relevant_filenames.txt"
    }
)
```

```
Traceback (most recent call last):
  File "/home/ubuntu/lightly-ai/lightly/test_deserialize.py", line 9, in <module>
    client.schedule_compute_worker_run(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 210, in schedule_compute_worker_run
    config_id = self.create_compute_worker_config(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 158, in create_compute_worker_config
    _validate_config(cfg=worker_config, obj=docker)
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 519, in _validate_config
    raise InvalidConfigurationError(error_msg)
lightly.api.api_workflow_compute_worker.InvalidConfigurationError: Option `relevant_filenames` does not exist! Did you mean `relevant_filenames_file`?
```
___ 
```python
client.schedule_compute_worker_run(
    worker_config={
        "corruptness_check": {
            "threshold": 1.0
        }
    }
)
```
```
Traceback (most recent call last):
  File "/home/ubuntu/lightly-ai/lightly/test_deserialize.py", line 9, in <module>
    client.schedule_compute_worker_run(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 210, in schedule_compute_worker_run
    config_id = self.create_compute_worker_config(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 158, in create_compute_worker_config
    _validate_config(cfg=worker_config, obj=docker)
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 521, in _validate_config
    _validate_config(item, getattr(obj, key))
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 519, in _validate_config
    raise InvalidConfigurationError(error_msg)
lightly.api.api_workflow_compute_worker.InvalidConfigurationError: Option `threshold` does not exist! Did you mean `corruption_threshold`?
```
___ 
```python
client.schedule_compute_worker_run(
    lightly_config={
        "modelx": {
            "name": "resnet-18"
        }
    }
)
```
```
Traceback (most recent call last):
  File "/home/ubuntu/lightly-ai/lightly/test_deserialize.py", line 9, in <module>
    client.schedule_compute_worker_run(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 210, in schedule_compute_worker_run
    config_id = self.create_compute_worker_config(
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 167, in create_compute_worker_config
    _validate_config(cfg=lightly_config, obj=lightly)
  File "/home/ubuntu/lightly-ai/lightly/lightly/api/api_workflow_compute_worker.py", line 519, in _validate_config
    raise InvalidConfigurationError(error_msg)
lightly.api.api_workflow_compute_worker.InvalidConfigurationError: Option `modelx` does not exist! Did you mean `model`
```

### CODE REVIEW

File: lightly/api/api_workflow_compute_worker.py [hunk 1]
```diff
@@ -489,3 +496,26 @@ def _snake_to_camel_case(snake: str) -> str:
     return components[0] + "".join(
         component.title() for component in components[1:]
     )
+
+
+def _validate_config(cfg: Optional[Dict[str, Any]], obj: Any) -> None:
+    """Validates that all keys in cfg are legit configuration options. """
```
Comment (michal-lightly) [inline lightly/api/api_workflow_compute_worker.py:502]: Please add more description about what this function is doing. It is quite non-trivial:
* I think it works only under certain assumptions on the type of `obj`
* It is recursive
* It suggests a best match

File: lightly/api/api_workflow_compute_worker.py [hunk 2]
```diff
@@ -489,3 +496,26 @@ def _snake_to_camel_case(snake: str) -> str:
...
+
+    if cfg is None:
+        return
+
+    for key, item in cfg.items():
+        if not hasattr(obj, key):
+            error_msg = f"Option `{key}` does not exist!"
+            if hasattr(type(obj), "swagger_types"):
```
Comment (michal-lightly) [inline lightly/api/api_workflow_compute_worker.py:510]: Suggestion: Add a comment that we suggest the closest possible match to the user for swagger classes. Can be here or in the function docstring.

Comment (michal-lightly) [reply lightly/api/api_workflow_compute_worker.py:510]: Optional: Maybe this should be an assert at the beginning of the function, or we can type `obj` with the union of allowed types. The motivation is that if we once switch away from swagger generator, the suggestion functionality silently stops working.

Edit: This would not work easily with recursive calls.

File: lightly/api/api_workflow_compute_worker.py [hunk 3]
```diff
@@ -489,3 +496,26 @@ def _snake_to_camel_case(snake: str) -> str:
...
+                closest_match = difflib.get_close_matches(
+                    word=key,
+                    possibilities=possible_options,
+                    n=1,
+                    cutoff=0.
+                )[0]
+                error_msg = f"{error_msg} Did you mean `{closest_match}`?"
+            raise InvalidConfigurationError(error_msg)
+        if isinstance(item, dict):
+            _validate_config(item, getattr(obj, key))
```
Comment (michal-lightly) [inline lightly/api/api_workflow_compute_worker.py:521]: Would this work also if `obj.key` is a dictionary?

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 4]
```diff
@@ -669,3 +667,45 @@ def test__snake_to_camel_case() -> None:
...
+    with pytest.raises(
+        InvalidConfigurationError,
+        match="Option `configx` does not exist!"
+    ):
+        _validate_config(
+            cfg={
+                "configx": "hello world",
+            },
+            obj=mock_obj
+        )
```
Comment (michal-lightly) [inline tests/api_workflow/test_api_workflow_compute_worker.py:711]: We should add more tests:
* A test on a real swagger class
* A test for a misspelling of a nested field

**Review by michal-lightly — CHANGES_REQUESTED** [review]
It seems to me that there are corner cases when this function might fail. It accepts any object as `obj`. If we want this to work just for swagger classes let's make it explicit.

File: lightly/api/api_workflow_compute_worker.py [hunk 5]
```diff
@@ -489,3 +496,26 @@ def _snake_to_camel_case(snake: str) -> str:
...
+
+def _validate_config(cfg: Optional[Dict[str, Any]], obj: Any) -> None:
+    """Validates that all keys in cfg are legit configuration options. """
+
+    if cfg is None:
+        return
+
+    for key, item in cfg.items():
+        if not hasattr(obj, key):
+            error_msg = f"Option `{key}` does not exist!"
```
Comment (michal-lightly) [inline lightly/api/api_workflow_compute_worker.py:509]: Minor: Let's use normal ticks instead of backticks

File: tests/api_workflow/test_api_workflow_compute_worker.py (hunk 4 above)
Comment (philippmwirth) [reply tests/api_workflow/test_api_workflow_compute_worker.py]: I replaced the mocked tests with real swagger classes.

File: lightly/api/api_workflow_compute_worker.py (hunk 2 above)
Comment (philippmwirth) [reply lightly/api/api_workflow_compute_worker.py]: I solved it by raising a type error when swagger types is missing.

File: tests/api_workflow/test_api_workflow_compute_worker.py [hunk 6]
```diff
@@ -669,3 +671,151 @@ def test__snake_to_camel_case() -> None:
...
+                    "num_workers": -1,
+                    "batch_size": 16,
+                    "shuffle": True,
+                },
+            },
+            obj=obj
+        )
+
+
+def test__validate_config__lightly_typo_nested(mocker: MockerFixture) -> None:
```
Comment (michal-lightly) [inline tests/api_workflow/test_api_workflow_compute_worker.py:788]: I think it is ok to test this (and the above) only once, but feel free to keep.
